File: src/main/java/com/dsa/problems/scaler/sort/merge_two_sorted_arrays.java
package com.dsa.problems.scaler.sort;

import java.util.ArrayList;
import java.util.Arrays;

public class merge_two_sorted_arrays {
  /**
   * Q. Merge Two Sorted Arrays
   *
   * Given two sorted integer arrays A and B, merge B and A as one sorted array and return it as an output.
   *
   * Note: A linear time complexity is expected and you should avoid use of any library function.
   */
  private static ArrayList<Integer> merge(ArrayList<Integer> A, ArrayList<Integer> B) {
    int i = 0, j = 0;
    ArrayList<Integer> temp = new ArrayList<>();
    while(i < A.size() && j < A.size()) {
      if(A.get(i) <= B.get(j)) {
        temp.add(A.get(i++));
      } else {
        temp.add(B.get(j++));
      }
    }
    while(i < A.size()) {
      temp.add(A.get(i++));
    }
    while(j < B.size()) {
      temp.add(B.get(j++));
    }

    return temp;
  }

  public static void main(String[] args) {
    System.out.println(merge(new ArrayList<>(Arrays.asList(2,3,5)), new ArrayList<>(Arrays.asList(1,4,6,7))));
  }
}

File: problems/scaler/sort/merge_two_sorted_arrays.js
/**
 * Q. Merge Two Sorted Arrays
 *
 * Given two sorted integer arrays A and B, merge B and A as one sorted array and return it as an output.
 *
 * Note: A linear time complexity is expected and you should avoid use of any library function.
 */
function soln(A, B) {
  let i = 0;
  let j = 0;
  let k = 0;
  const n = A.length;
  const m = B.length;
  const rslt = new Array(n + m).fill(0);
  while (i < n && j < m) {
    if (A[i] <= B[j]) {
      rslt[k] = A[i];
      i++;
      k++;
    } else {
      rslt[k] = B[j];
      k++;
      j++;
    }
  }
  while (i < n) {
    rslt[k] = A[i];
    k++;
    i++;
  }
  while (j < m) {
    rslt[k] = B[j];
    k++;
    j++;
  }

  return rslt;
}

console.log(soln([4, 7, 9], [2, 11, 19]));
